### DISCUSSION TITLE
Launcher icons

### DISCUSSION THREAD

**Discussion by Taitava** [Taitava/obsidian-shellcommands#25] [Ideas]
## Update on 2022-03-01
There are **three** different icons related subjects here:
1. Launcher icons in the left side bar of Obsidian. This is what this discussion originally discussed about.
2. Menu icons for shell commands (related to _events_, discussed below).
3. Changing some icons in the settings UI of this plugin. I'm not sure if I have discussed about this somewhere else or not. Currently some of the icons I use in the settings panels are not so clear about what they do, but luckily there are tooltips clarifying what they mean.

**Each of the three subjects should have their _own, separate issues_ created during implementation.**

## Update on 2022-01-01

Now that _events_ are coming (#60 and #123), I'm planning to warm up this idea of icons again. _Events_ will make it possible to insert shell commands in three different context menus:
 - Context menu on a file in the navigation pane,
 - Context menu on a folder in the navigation pane, and
 - Context menu in an editor pane.

At first, shell commands will not have any icons in the menus, but I'd like to add an ability for a user to pick an icon that will appear in the menus.

---
Original post:

Currently commands can be executed via the command palette or hotkeys. I'm thinking about a third option, you could define certain commands to have an icon button in the left side bar.

Options that you can choose:
 - Which commands will have an icon. Not all commands need an icon. The ones that you execute all the time are probably commands that you have assigned a hotkey, so they do not need an icon. But rarely used ones can benefit from an icon: you do not have to reserve a hotkey for them, and you do not need to search them from the command palette.
 - Choose an icon from a list of icons. These icons come from Obsidian (actually, probably some library used by Obsidian?), and I'm not planning to add support for users to add their own icons. The list of available icons in Obsidian is quite good.
 - Choose if the icon will be placed to top left (near "open Markdown importer" icon) or bottom left (near "Open another vault" icon).
 - Maybe the order of shell command icons could be decided too? I'm not sure yet.

A launcher icon has one downside. I can add a tooltip to it so that if you hover over the icon with mouse, it will show a text with the command's alias (or the command itself if no alias is defined), but the tooltip does not support [variable previewing](https://github.com/Taitava/obsidian-shellcommands/issues/10). This is simply because tooltips are generated at the same time when the button is generated, so making variables previewable in the tooltips would have a risk of stale values being shown in the tooltips. But now that I think of it, maybe there is a way to hook into the mouse hover event and update the tooltip when the mouse is on an icon. I'll have to think about it.

**Comment by FelipeRearden** [comment]
Hello @Taitava !

As a source of information about this idea, take a look at these plug-ins:

[cMenu](https://github.com/chetachiezikeuzor/cMenu-Plugin) from Chetachi

[Customizable Sidebar](https://github.com/phibr0/obsidian-customizable-sidebar) from Philbro

I think you gonna like it :)

Have a great day!!!!

**Comment by Taitava** [comment]
Excellent, @FelipeRearden !

I think I'll postpone this idea. I should add a new section to README.me: *Benefits from other plugins* . There would be a list of plugins that could be used in combination with this one (*cMenu* and *Customizable Sidebar* like you mentioned) and plugins providing alternative functionality ([Text Expander](https://forum.obsidian.md/t/text-expander-plugin/11524), allows you to write and execute shell commands in notes).

The idea of implementing icon functionality in this plugin can still be beneficial in the future. It could simplify things for the user: if they only need to add icons related to shell commands, they would not need to install another plugin. On the UI side, this would add one form element to the *cog modal*/whatever it is, and the user would not have to go to another setting view. This would be a similar icon as in *Customizable Sidebar*, not like in *cMenu*. But for now, the ability to use *Customisable Sidebar* with this plugin is good, and we can concentrate on other ideas. 🙂 

*cMenu* has a different idea than I originally had, but it's clever and maybe some shell commands that utilize `{{selection}}` could benefit from it. It's nice to have this kind of companion plugin, too! 🙂

**Comment by FelipeRearden** [comment]
This is a great idea @Taitava !!!

So far, my _Benefits from other plugins _ are:
- [QuickAdd](https://github.com/chhoumann/quickadd)
- [Obsidian Advanced URI](https://github.com/Vinzent03/obsidian-advanced-uri)

I let you know when I found a new companion plugin to this list 

I didn’t know about Text Expander - I’m gonna take a look. Thanks for sharing!

> The idea of implementing icon functionality in this plugin can still be beneficial in the future. It could simplify things for the user.

100% agree with you. It is just a matter of time. 

Have a great day!

**Comment by Taitava** [comment]
> I didn’t know about Text Expander - I’m gonna take a look. Thanks for sharing!

I haven't actually used it, but I found it when searching for other plugins that might offer shell command functionality. It's quite rare for me to test new plugins. Maybe I should do it more often.

**Reply by FelipeRearden** [reply]
My habit is to always search for new plugins even when I don’t have a usage in the present.

Because I know that this insight will trigger in the future. Obsidian has some new features everyday that could change my workflow every time.

**Comment by Taitava** [comment]
> Obsidian Advanced URI

You know this plugin far better than I do. Can you write an example advanced Obsidian URI that is very useful to you? I know we have already *obsidian://advanced-uri?vault=my%20vault&workspace=main* in the README.md's *Usage examples* section. Does something more come to your mind? I could have a new example in the *Benefits from other plugins* section.

**Reply by FelipeRearden** [reply]
Sure! 

I was using this one before you show to me `>>` : Append clipboard to a specific file:

`open “obsidian://advanced-uri?vault=Vault_name&filepath=Filename.md&data={{clipboard}}&mode=append”`

**Reply by Taitava** [reply]
Thanks, added!

**Reply by FelipeRearden** [reply]
One more for our list!!!!!!!! 

Obsidian URI:
`open “obsidian://open?vault=Notes”` -> to open a different Obsisian vault 

[Source: a question in the forum](https://forum.obsidian.md/t/opening-another-vault-with-keyboard-only/24359)

**Reply by Taitava** [reply]
Thank you @FelipeRearden for mentioning this plugin in the forums!

Actually, we already have this example in the readme: `start "" "obsidian://open?vault=my%20vault"`

It's just in Windows format. But thanks for suggesting anyway! 🙂

**Reply by FelipeRearden** [reply]
Oops. Sorry my bad :)

**Comment by Taitava** [comment]
I created the *Benefits from other plugins* section a while ago and now I also added info about *QuickAdd*.

**Reply by FelipeRearden** [reply]
Great!

**Reply by FelipeRearden** [reply]
I just wrote the new section and its perfect!!!!

And thanks for the personal reference!!!!!

**Reply by Taitava** [reply]
You are welcome! 🙂

**Comment by Taitava** [comment]
# Events: Add icons to context menus
I updated the first post with information about adding icons to context menus some time later.

**Comment by Taitava** [comment]
# Obsidian supports more icons now

From Obsidian forum, topic [Obsidian Release v0.13.27 (Insider build)](https://forum.obsidian.md/t/obsidian-release-v0-13-27-insider-build/33223):
> - Obsidian’s icon library now supports all icons from `lucide`, a currently maintained fork of feather-icons with 547 supported icons. No need to import feather-icons anymore. You can browse them at https://lucide.dev/

This is good news so I don't have to import any external library when I'll switch to a larger set of icons. I'm not implementing these icon features/changes anytime soon, so even though Obsidian `0.13.27` is currently an insider build, it will probably be available to everybody by the time I'll get my hands on doing this icon stuff.
